File: 2488-divide-intervals-into-minimum-number-of-groups/2488-divide-intervals-into-minimum-number-of-groups.py
class Solution:
    def minGroups(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x: x[0])
        heap = []
        min_num_of_machines_required = 0

        for interval in intervals:
            start_time, end_time = interval
            while heap and heap[0] < start_time:
                heapq.heappop(heap)
            heapq.heappush(heap, end_time)
            min_num_of_machines_required = max(min_num_of_machines_required, len(heap))

        return min_num_of_machines_required
        

File: 2488-divide-intervals-into-minimum-number-of-groups/2488-divide-intervals-into-minimum-number-of-groups.java
class Solution {
    public int minGroups(int[][] intervals) {
        Arrays.sort(intervals, (a,b)->a[0]-b[0]);
        PriorityQueue<Integer>pq=new PriorityQueue<>();
        int minGroups = 0;
        for(int[] interval:intervals){
            int start =interval[0];
            int end = interval[1];
            while(!pq.isEmpty() && pq.peek() < start){
                pq.poll();
            }
            pq.add(end);
            minGroups = Math.max(minGroups, pq.size());
        }
        return minGroups;
       
    }
}